Type check KNX integration __init__ and knx_entity

farmio opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Typecheck KNX integration. Step by step...

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #48006
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

farmio committed:
typecheck knx __init__

farmio committed:
typecheck knx knx_entitiy

farmio commented on homeassistant/components/knx/__init__.py:
```diff
@@ -398,16 +406,16 @@ def register_callback(self) -> TelegramQueue.Callback:
             map(AddressFilter, self.config[DOMAIN][CONF_KNX_EVENT_FILTER])
         )
-        return self.xknx.telegram_queue.register_telegram_received_cb(
+        return self.xknx.telegram_queue.register_telegram_received_cb(  # type: ignore[no-any-return]
```
I have no idea why mypy handles this as `Any`. It returns a TelegramQueue.Callback in xknx code 
https://github.com/XKNX/xknx/blob/6a9be156132bac5233a0a5000e8ea953369d642d/xknx/core/telegram_queue.py#L80
I have tried changeing it to a forward reference from using only `-> Callback:` previously (xknx 0.17.2) but this didn't change anything.

frenck replied:
🤷

farmio commented on homeassistant/components/knx/__init__.py:
```diff
@@ -445,27 +453,28 @@ async def service_exposure_register_modify(self, call):
             )
             replaced_exposure.shutdown()
-        exposure = create_knx_exposure(self.hass, self.xknx, call.data)
+        exposure = create_knx_exposure(self.hass, self.xknx, call.data)  # type: ignore[arg-type]
```
`call` is a `homeassistant.helpers.typing.ServiceCallType`. 
Why is `call.data` not a `homeassistant.helpers.typing.ServiceDataType`, but a `MappingProxyType[Any, Any]` here?
`create_knx_exposure` is expecting a `ConfigType` (`Dict[str, Any]`) as 3rd argument. Should I change this for  a `Union[ConfigType,  MappingProxyType[Any, Any]]` ?

frenck replied:
Use the `ServiceCall` from core intead.

farmio replied:
Result is the same.
https://github.com/home-assistant/core/blob/88b5eff726155d44064e4c954e00ef8e326c2046/homeassistant/helpers/typing.py#L13

frenck commented on homeassistant/components/knx/__init__.py:
```diff
@@ -296,19 +302,19 @@ async def reload_service_handler(service_call: ServiceCallType) -> None:
 
-    def __init__(self, hass, config):
-        """Initialize of KNX module."""
+    def __init__(self, hass: HomeAssistantType, config: ConfigType):
```
```suggestion
    def __init__(self, hass: HomeAssistantType, config: ConfigType) -> None:
```

frenck commented on homeassistant/components/knx/__init__.py:
```diff
@@ -428,9 +436,9 @@ async def service_event_register_modify(self, call):
                     )
 
-    async def service_exposure_register_modify(self, call):
+    async def service_exposure_register_modify(self, call: ServiceCallType) -> None:
```
```suggestion
    async def service_exposure_register_modify(self, call: ServiceCall) -> None:
```

frenck commented on homeassistant/components/knx/__init__.py:
```diff
@@ -445,27 +453,28 @@ async def service_exposure_register_modify(self, call):
         )
 
-    async def service_send_to_knx_bus(self, call):
+    async def service_send_to_knx_bus(self, call: ServiceCallType) -> None:
```
```suggestion
    async def service_send_to_knx_bus(self, call: ServiceCall) -> None:
```

frenck commented on homeassistant/components/knx/__init__.py:
```diff
@@ -483,9 +492,9 @@ async def service_send_to_knx_bus(self, call):
             await self.xknx.telegrams.put(telegram)
 
-    async def service_read_to_knx_bus(self, call):
+    async def service_read_to_knx_bus(self, call: ServiceCallType) -> None:
```
```suggestion
    async def service_read_to_knx_bus(self, call: ServiceCall) -> None:
```

farmio committed:
Merge remote-tracking branch 'upstream/dev' into knx-mypy-init

farmio committed:
return type for init

farmio committed:
enable type checking for knx

farmio commented:
Since this is the last missing typing PR for KNX I have added it to `setup.cfg`. Also added a blank test so xknx will be added as a test dependency and mypy can load its annotations.

farmio committed:
Merge remote-tracking branch 'upstream/dev' into knx-mypy-init

farmio committed:
update for xknx 0.17.4

farmio committed:
use core.ServiceCall instead of typing.ServiceCallType

frenck commented on tests/components/knx/__init__.py:
```diff
@@ -0,0 +1,3 @@
+"""Tests for the KNX integration."""
```
This is a workaround, we should not do this until there are actually tests

farmio replied:
First tests are on their way anyway. https://github.com/home-assistant/core/pull/48311
I just wanted to avoid awkward conflicts for other PRs that break mypy checks without getting notified (because of `ignore_missing_imports = True`). 
I'll remove it again...

frenck replied:
Well, easy solution: Merge #48311 first, rebase this one.

farmio replied:
Sure, go ahead.

farmio replied:
So tests are available now. I think we are set.

farmio committed:
Merge remote-tracking branch 'upstream/dev' into knx-mypy-init

farmio commented:
So tests are available now. I think we are set.

frenck reviewed (approved):
🎉  Thanks, @farmio 👍
